blueprints/app/files/package.json

```diff
@@ -17,7 +17,8 @@
     "lint:hbs": "ember-template-lint .",
     "lint:hbs:fix": "ember-template-lint . --fix",
     "lint:js": "eslint . --cache",
-    "lint:js:fix": "eslint . --fix",
+    "lint:js:fix": "eslint . --fix",<% if (typescript) { %>
+    "lint:ts": "tsc --noEmit",<% } %>
```

Comment:
The RFC also doesn't say anything about this. But I *believe* something like this should be added!?

Reply:
Something longer it, yes, but it should absolutely not be named as a lint. I know a bunch of addons do treat it this way, and some people choose to think of it this way, but from a pedagogical perspective it is not how we (specifically the Typed Ember team) want to encourage people to think of TypeScript. Type errors are a different class of errors than lint errors. They aren’t opinions or best practices, however string: they are *correctness issues*, and so deserve a different treatment.

I would suggest something like `type-check` or, if we wanted to generalize, introduce a `check:` namespace which could include `check:types` and `check:lint:*`.

Reply:
I see your point, and you are certainly not wrong here. 

However the solution here seems like the most pragmatic, as putting it under `lint:*` makes the main `lint` script also run the TS check, alongside ESLint and template lint (by means of the `npm-run-all` setup). Putting it under a different name(space) would require users to account for it in their CI setup, and us to update the CI-parts of the blueprints. 

Or we define `lint` as something like `"lint": "npm-run-all --aggregate-output --continue-on-error --parallel \"lint:!(fix)\" && yarn type-check"`, but this doesn't make things much better with regard to your argument of not confusing type checking with linting.

Reply:
I have super mixed feelings on this one. In a vacuum designing this from scratch, I like the idea of a `check:` namespace, but I suspect that would be a complicated change to make today. I think there would be a fair amount of friction in migrating, both in terms of existing users' muscle memory and custom scripts, and how `lint:fix` is enshrined in `ember-cli` as the "run this after generating new files to clean them up" script.

I agree that Including `type-check` (or `typecheck`, depending on what color we paint that shed) in the `lint` script isn't ideal, but it may be an ok pragmatic compromise.

Reply:
> type errors are a different class of errors than lint errors. They aren’t opinions or best practices, however strong: they are correctness issues, and so deserve a different treatment.

I don't think this is a helpful framing because lots of lint rules are *also* correctness issues, up to and including "instant runtime failure".

Reply:
I think we should zoom in on the actual pedagogical goal here, which is probably something like "helping people understand that some checks are pretty important for correctness while others are just opinions so they don't turn off the important ones".

But trying to draw a hard line between tsc and any other static checks that are happening is not an effective solution to that problem. There are eslint and template-lint rules that are *also* about correctness just as much as typescript is, so if we're trying to teach people what's actually important, the boundary between important and opinion should include some subset of the eslint and template-lint rules too.

I also think you're never going to get everyone to agree that tsc is not a linter, so unless you want to keep having that argument forever you can just accept that it sure looks very linter-esque, especially when it's running `--noEmit` and literally has the exact same developer interactions that eslint does ("During CI it does static analysis to tell me about problems and in my editor it makes red squiggles.").

Finally, I would argue that typescript-the-project thinks about itself as a linter. That's why `noEmitOnError` defaults to false. It's **normal and expected** during active development **to run and debug code that still has type errors**, and that is central to the design of typescript.

Reply:
> it does static analysis to tell me about problems and in my editor it makes red squiggles.

Yes, that pretty much sums up how I personally understand the term linting. For me it was always about *how* it does its job (static analysis, i.e. basically "looking" at code instead of running it, which is what tests are about).

Similarly, the [definition of linting on Wikipedia](https://en.wikipedia.org/wiki/Lint_(software)) emphasizes static analysis, but does not really limit the scope of what kind of code problems it cares about: (see "programming errors" which I understand as *correctness* type of problems)

> Lint, or a linter, is a [static code analysis](https://en.wikipedia.org/wiki/Static_program_analysis) tool used to flag programming errors, [bugs](https://en.wikipedia.org/wiki/Software_bug), stylistic errors and suspicious constructs.

Reply:
I take the point that ESLint and ember-template-lint do both catch actual *correctness* issues; I was wrong when I distinguished along that axis. That being said, I still think there is a meaningful difference here—types, lints, formal methods, preconditions-and-postconditions, and more all exist on the spectrum of “static code analysis tools for proving correctness.” That doesn’t mean they’re the same things! However, I’m *much* more interested in seeing this move forward than in getting lost in the weeds about that detail. I am fine with going with `lint:types` for now, and we can think about whether we want to move the ecosystem to the idea of `check:*` or such at some later point.

I'm suggesting `lint:types` rather than `lint:ts` because I think `lint:ts` might imply "do ESLint or similar on TS files".

```suggestion
    "lint:types": "tsc --noEmit",<% } %>
```

Reply:
Updated!